Add flag to disable legacy services

brb opened the pull request:
This PR introduces the `--enable-legacy-services=<bool>` flag  (`true` by default) which can enable or disable the legacy services.

There are other changes in this PR worth mentioning:

* We do not calculate the service weights anymore, as the weighted services are disabled in the datapath code and disabling the calculation makes my changes a bit more simple.
* From now on, services v2 are considered the source of truth.
* Remove backend and release its ID during the restore if there is no service pointing to it.
* Remove service v2 if there is no backing legacy equivalent during the restore.
 
I have tested the PR manually. Also, I changed the default of the flag to `false` and then ran the CI tests. Anyway, the tests are coming in my next PR.

Please review each commit separately.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7639)
<!-- Reviewable:end -->

brb committed:
bpf: Add Map.DumpWithCallbackIfExists method

This method is going to be used by the `cilium service list`
handler to dump and parse svc v2 entries.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
bpf: Add Map.UnpinIfExists method

This method is going to be used for removal of the legacy svc
related maps which might not exist during the removal.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
cli: Use svc v2 maps when listing

This commit changes the source maps of the `cilium bpf lb list`
cmd - instead of the legacy svc BPF maps, the v2 BPF are used.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
loadbalancer: Sort backends by ID when listing

When listing services with `cilium service list`, the order of backends
might not reflect the actual order of service values in the svc v2
BPF maps.

After the cilium agent has been restarted, it restores the loadbalancer
cache from the BPF maps. So, the service value entries might be
in a different order than it used to be before the restart. This is
fine, but it makes some integrations test to fail which expect
service values to be in the same order as before the restart.

This commit ensures that the service values are listed always in the
same order (i.e. sorted by the backend ID).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon,bpf: Add --enable-legacy-services flags

The flag (by default set to "true") enables or disables the legacy
services. This can be useful to avoid unnecessary memory usage and
to save some CPU cycles when doing a service lookup in the datapath
for those deployments which are going to use Cilium v1.5 from the
beginning, i.e. won't do an upgrade from v1.4 to v1.5.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap,daemon: Make removal of lbmap cache more explict

This commit introduces the `lbmap.DeleteServiceCache` function
which is responsible for the removal of the lbmap cache svc
entry.

Also, the removal is moved from a function which was called
when a legacy svc has been removed to a function which
manages the removal of both types of services, so that we
still remove the cache entry even when the legacy services
are disabled.

Once we stop support the legacy services, we can move the removal
to the `lbmap.DeleteServiceV2` function.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Exclude master service earlier in dump function

As we do not use master service to extract a service ID [1],
we can ignore parsing of the master service earlier which
makes the `lbmap.DumpServiceMapsToUserspaceV2` function
smaller.

[1]: https://github.com/cilium/cilium/pull/7540#discussion_r272913219

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Update revNAT table from v2 routines

This commit moves the creation and update of rev nat tables
from the update function of the legacy services to the update
function of the v2 services.

This is needed, because we will optionally disable the legacy
services.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon,lbmap: Do not update legacy svc if they are disabled

This commit disables updating of the legacy services if they
are disabled (`--enable-legacy-services=false`).

Two other changes:

- Restoring from the legacy services is disabled if the legacy
  services are disabled.
- As we do not support the weighted services, the calculation of
  the service endpoint weights is disabled to simplify the code.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Remove legacy svc BPF maps if they are disabled

This commit tries to remove the legacy svc related maps if the legacy
services are disabled.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Use v2 services when syncing with k8s

The syncing (`syncLBMapsWithK8s`) happens after the v2 services
have been created from the legacy ones, so we can use the v2
for comparing against the k8s ones when syncing.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -95,8 +115,13 @@ func (s *LBSVC) GetModel() *models.Service {
-		spec.BackendAddresses[i] = be.GetBackendModel()
+		placements[i] = backendPlacement{pos: i, id: be.ID}
+	}
+	sort.Sort(placements)
```
Instead of implementing the sort interface, you have the https://golang.org/pkg/sort/#Slice

aanm reviewed (commented):
Looks good overall. The only missing part left is to disable the legacy services by default in k8s configmap

brb committed:
logfields: Fix BackendID logfield value

The field is used when logging a single instance of backend.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Fix BackendAddrID of IPv6 backend

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add BackendAddrID.IsIPv6 method

This method detects (in a dirty way) whether a backend identified with
the given addrID is of the ipv6 type.

It will be used when removing orphan backends from the BPF maps.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon,lbmap: Remove orphan v2 services

It's possible to end up in a situation when a v2 service does
not have an equivalent legacy service. Consider the following:

1. cilium-agent upgraded to >= v1.5.
2. cilium-agent downgraded to < v1.5.
3. The service is removed (only from the legacy BPF map).
4. cilium-agent upgraded to >= 1.5.

If this happen, in the 4. step v2 service won't be restored from
legacy. To avoid it, we remove orphan services after we have migrated
legacy services to v2.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon,lbmap: Remove orphan backends

This commit removes orphan backends (i.e. backends which are not used
by any service) and releases their IDs.

ianvernon commented on cilium/cmd/bpf_lb_list.go:
```diff
@@ -15,8 +15,12 @@
 package cmd
 
 import (
+	"fmt"
```
I think we should be able to dump both using some type of flag. Defaulting to `v2` is a good idea, but we should be able to introspect `v1` as well if some option, `--v1` is provided to this command as well. This can be done as a follow-up PR to avoid further churn.

brb replied:
Noted. Sure, I will do it in a follow-up PR.
